public void testRereplicate3() throws Exception {
    List<DatanodeDescriptor> chosenNodes = new ArrayList<DatanodeDescriptor>();
    chosenNodes.add(dataNodes[0]);
    chosenNodes.add(dataNodes[2]);
    DatanodeDescriptor[] targets;
    targets = replicator.chooseTarget(0, dataNodes[0], chosenNodes, null, BLOCK_SIZE);
    assertEquals(targets.length, 0);
    targets = replicator.chooseTarget(1, dataNodes[0], chosenNodes, null, BLOCK_SIZE);
    assertEquals(targets.length, 1);
    assertTrue(cluster.isOnSameRack(dataNodes[0], targets[0]));
    assertFalse(cluster.isOnSameRack(dataNodes[2], targets[0]));
    targets = replicator.chooseTarget(1, dataNodes[2], chosenNodes, null, BLOCK_SIZE);
    assertEquals(targets.length, 1);
    assertTrue(cluster.isOnSameRack(dataNodes[2], targets[0]));
    assertFalse(cluster.isOnSameRack(dataNodes[0], targets[0]));
    targets = replicator.chooseTarget(2, dataNodes[0], chosenNodes, null, BLOCK_SIZE);
    assertEquals(targets.length, 2);
    assertTrue(cluster.isOnSameRack(dataNodes[0], targets[0]));
    targets = replicator.chooseTarget(2, dataNodes[2], chosenNodes, null, BLOCK_SIZE);
    assertEquals(targets.length, 2);
    assertTrue(cluster.isOnSameRack(dataNodes[2], targets[0]));
}
